core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -108,8 +108,59 @@ class TransactionStateManager(brokerId: Int,
     }
   }
 
-  def enablePidExpiration() {
-    // TODO: add producer id expiration logic
+  def enableTransactionalIdExpiration() {
+    scheduler.schedule("transactionalId-expiration", () => {
+      inLock(stateLock) {
+        val now = time.milliseconds()
+
+        val transactionalIdByPartition: Map[Int, mutable.Iterable[String]] = transactionMetadataCache.flatMap { case (_, entry) =>
+          entry.metadataPerTransactionalId.filter { case (_, txnMetadata) =>
+            txnMetadata.txnLastUpdateTimestamp <= now - config.transactionalIdExpirationMs
+          }.map { case (transactionalId, _) =>
+            transactionalId
+          }
+        }.groupBy { transactionalId =>
+          partitionFor(transactionalId)
+        }
+
+        val recordsPerPartition = transactionalIdByPartition
```

Comment:
I am concerning if the following can happen:

1) tc decides to expire a txnId, and have written the tombstone.
2) an `addPartitions` is received from the txnId, and is accepted and handled since no checks will fail.
3) tombstone is replicated, the cache is removed from the cache.

Should we guard the object if we have decided to remove the txnId? For example, we can add a new state, say `Dead` in the `TransactionState`, and set its pending state to that before we decided to expire and append to local log. And when handling `addPartition` and `endTxn`, if its pending state is `Dead` return concurrent txn to let it retry later, and when handling `initPid` and pending state is `Dead`, bump up the producer epoch as we did; then on the `removeFromCacheCallback`, check the pending state / coordinator epoch / state to make sure it is `Dead / same as the old epoch / same as the old state` and only remove the cache if it passes.